include/oneapi/dpl/internal/function.h

```diff
@@ -63,33 +64,153 @@ using oneapi::dpl::__par_backend_hetero::__internal::__buffer;
 counting_iterator<T>
-get_access(Policy, counting_iterator<T> i)
+get_access_at(Policy, counting_iterator<T> i, size_t offset)
 {
-    return i;
+    return i + offset;
 }
 
 template <sycl::access::mode Mode, typename Policy, typename T>
 T*
-get_access(const Policy& policy, T* ptr)
+get_access_at(const Policy& policy, T* ptr, size_t offset)
 {
     assert(sycl::get_pointer_type(ptr, policy.queue().get_context()) == sycl::usm::alloc::shared);
-    return ptr;
+    return ptr + offset;
+}
+
+template <typename ValueType, typename Policy, typename Iterator>
+void
+get_data_at(const Policy& p, Iterator i, size_t index, ValueType& val)
+{
+    val = get_access_at<sycl::access::mode::read>(p, i, index)[0];
 }
+
+template <typename ValueType, typename Policy, typename T>
+void
+get_data_at(const Policy& policy, T* ptr, size_t index, ValueType& val)
+{
+    sycl::queue q = policy.queue();
+
+    switch (sycl::get_pointer_type(ptr, q.get_context()))
+    {
+    case sycl::usm::alloc::host:
+    case sycl::usm::alloc::shared:
+        val = ptr[index];
+        break;
+
+    case sycl::usm::alloc::device:
+        q.copy(ptr + index, &val, 1);
+        q.wait();
+        break;
+
+    case sycl::usm::alloc::unknown:
+        assert(!"Unknown pointer type");
+    }
+}
+
+template <typename ValueType, typename Policy, typename Iterator>
+void
+set_data_at(const Policy& p, Iterator i, size_t index, ValueType val)
+{
+    get_access_at<sycl::access::mode::write>(p, i, index)[0] = val;
+}
+
+template <typename ValueType, typename Policy, typename T>
+void
+set_data_at(const Policy& policy, T* ptr, size_t index, ValueType val)
+{
+    sycl::queue q = policy.queue();
+
+    switch (sycl::get_pointer_type(ptr, q.get_context()))
+    {
+    case sycl::usm::alloc::host:
+    case sycl::usm::alloc::shared:
+        ptr[index] = val;
+        break;
+
+    case sycl::usm::alloc::device:
+        q.copy(&val, ptr + index, 1);
+        q.wait();
+        break;
+
+    case sycl::usm::alloc::unknown:
+        assert(!"Unknown pointer type");
+    }
+}
+
+template <typename Policy, typename IteratorSrc, typename IteratorDest>
+void
+copy_data_to(const Policy& p, IteratorSrc itSrc, size_t indexSrc, IteratorDest itDest, size_t indexDest)
+{
+    auto accRead = get_access_at<sycl::access::mode::read>(p, itSrc, indexSrc);
+    auto accWrite = get_access_at<sycl::access::mode::write>(p, itDest, indexDest);
+
+    accWrite[0] = accRead[0];
+}
+
+template <typename Policy, typename IteratorSrc, typename T>
+void
+copy_data_to(const Policy& p, IteratorSrc itSrc, size_t indexSrc, T* ptrDest, size_t indexDest)
+{
+    auto accRead = get_access_at<sycl::access::mode::read>(p, itSrc, indexSrc);
+    set_data_at(p, ptrDest, indexDest, accRead[0]);
+}
+
+template <typename Policy, typename T, typename IteratorDest>
+void
+copy_data_to(const Policy& p, T* ptrSrc, size_t indexSrc, IteratorDest itDest, size_t indexDest)
+{
+    T val;
```

Comment:
I think we can copy data directly, w/o a host temporary variable `val`.

Reply:
Fixed